upf/dnf.py

```diff
@@ -0,0 +1,179 @@
+                    # while the arguments can be expanded in this
+                    # iteration.
+                    stack.append((p, new_e, True))
+                    stack.append((not p, e.arg(0), False))
+                    stack.append((p, e.arg(1), False))
+                elif e.is_iff():
+                    na1 = self.manager.Not(e.arg(0))
+                    na2 = self.manager.Not(e.arg(1))
+                    e1 = self.manager.And(e.arg(0), e.arg(1))
+                    e2 = self.manager.And(na1, na2)
+                    new_e = self.manager.Or(e1, e2)
+                    #stack.append((p, new_e, False)) would be enough.
+                    # but this requires more iterations on the stack
+                    # while the arguments can be expanded in this
+                    # iteration.
+                    stack.append((p, new_e, True))
+                    stack.append((p, e1, True))
+                    stack.append((p, e.arg(0), False))
+                    stack.append((p, e.arg(1), False))
+                    stack.append((p, e2, True))
+                    stack.append((not p, e.arg(0), False))
+                    stack.append((not p, e.arg(1), False))
+                else:
+                    if p:
+                        solved.append(e)
+                    else:
+                        solved.append(self.manager.Not(e))
+        return solved.pop()
+
+
+
+
+class Dnf(DagWalker):
+    """Class used to transform a logic expression into the equivalent
+    Disjunctive Normal Form expression.
+
+    This is done first by transforming the expression into a NNF expression,
+    and then every And and Or are propagated to be a unique equivalent Or of
+    Ands or Atomic expressions, where 'atomic expressions' could also be a
+    Not of an atomic expression.
+    """
+    def __init__(self, env: 'upf.environment.Environment'):
+        DagWalker.__init__(self, True)
+        self.env = env
+        self.manager = env.expression_manager
+        self._nnf = Nnf(self.env)
+
+    def get_dnf_expression(self, expression: FNode) -> FNode:
+        """Function used to transform a logic expression into the equivalent
+        Disjunctive Normal Form expression.
+
+        This is done first by transforming the expression into a NNF expression,
+        and then every And and Or are propagated to be a unique equivalent Or of
+        Ands or Atomic expressions, where 'atomic expressions' could also be a
+        Not of an atomic expression.
+
+        For example, the form: !(a => (b && c)) becomes:
+        a && (!b || !c), in NNF form, and then:
+        (a && !b) || (a && !c), therefore a DNF expression."""
+        nnf_exp = self._nnf.get_nnf_expression(expression)
+        tuples = self.walk(nnf_exp)
+        and_list: List[FNode] = []
+        for and_args in tuples:
+            and_list.append(self.manager.And(and_args))
+        return self.manager.Or(and_list)
```

Comment:
An iterator is an Iterable and both `And` and `Or` accept an Iterable as first argument, so you can simplify the code:
```suggestion
        return self.manager.Or(self.manager.And(and_args) for and_args in tuples)
```